segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/DirectMemoryCache.java

```diff
@@ -399,15 +403,23 @@ private DirectMemoryBuffer getNextAvailableBuffer() {
 
             // If we get here, there are no available buffers and we have allocated all the buffers we could. Notify
             // any upstream listeners to attempt a cleanup (if possible).
-        } while (++attempts <= MAX_CLEANUP_ATTEMPTS && tryCleanup());
+            attempts++;
+            tryCleanup(attempts);
+        }
 
         // Unable to reuse any existing buffer or find a new one to allocate and upstream code could not free up data.
         throw new CacheFullException(String.format("%s full: %s.", DirectMemoryCache.class.getSimpleName(), getState()));
     }
 
-    private boolean tryCleanup() {
+    private void tryCleanup(int attempts) {
         val c = this.tryCleanup.get();
-        return c != null && c.get();
+        if (c != null && !c.get()) {
+            // Unable to clean up the cache. Wait a bit, then try again.
+            int sleepMillis = attempts * this.retryDelayBaseMillis.get();
```

Comment:
RetryPolicy in common is built to do this.

Reply:
That was my first choice, however `Retry` only retries on exceptions. Here I retry based on the response from a method.